### DISCUSSION TITLE
Is the `imagesdir` set wrong for AsciiDoc?

### DISCUSSION THREAD

**Discussion by guenhter** [docToolchain/docToolchain#1396] [Q&A]
Hi,

I have this structure

```
images1/foo.png
images2/bar.png
sub/some.adoc
docToolchainConfig.groovy
```

The adoc looks like this:

```adoc
= Simple Test

image::foo.png[]
image::bar.png[]
```

and the config is:

```groovy
inputPath = ''

inputFiles = [
        [file: 'sub/some.adoc',  formats: ['html','pdf']],
]

imageDirs = [
	"./images1/",
	"./images2/",
]
...
```

but when running this, the images are not found:

```
Converting /project/sub/some.adoc
Apr 24, 2024 11:08:46 AM uri:classloader:/gems/asciidoctor-pdf-2.3.4/lib/asciidoctor/pdf/converter.rb convert_image
WARNING: image to embed not found or not readable: /project/sub/images/foo.png
WARNING: image to embed not found or not readable: /project/sub/images/bar.png
Exception in thread "main" org.asciidoctor.gradle.remote.AsciidoctorRemoteExecutionException: ERROR: The following messages from AsciidoctorJ are treated as errors:
image to embed not found or not readable: /project/sub/images/foo.png
```

It seems like, that the `imagesDir` folder is set wrong by the doctoolchain, am I right?

**Comment by ponchofiesta** [comment]
I have a similar problem.

```
publish build/html5/src/asciidoc/doc/product/Anwenderhandbuch/Anwenderhandbuch.html
Publish /project/build/html5/src/asciidoc/doc/product/Anwenderhandbuch/Anwenderhandbuch.html to ~ajhmf7y at https://devstack.vwgroup.com/confluence/rest/api/ ...
GAF-NX Anwenderhandbuch GAF-NX Anwenderhandbuch
    image: /project/build/html5/src/asciidoc/doc/product/Anwenderhandbuch/../images/internal-badge.png
    image: /project/build/html5/src/asciidoc/doc/product/Anwenderhandbuch/images/diagram20240813a.png
```
The first image path is missing one step up (".."). In the Anwenderhandbuch.adoc I set "../../images/internal-badge.png"
The second one is a PlantUML diagram which is working.

**Comment by rdmueller** [comment]
sorry for this trouble.

the `imagesdir` is handled in some unconventional way in asciidoc, because it is always relative to the document which is currently rendered. Not relative to the document in which it is referenced.

So, if you have a document

    src/docs/main.adoc

which includes

    src/docs/chapters/chapter1.ado

which references

    image::foo.png

the image is expected to be found in `src/docs/.` when `src/docs/main.adoc` is rendered and in `src/docs/chapters/.` when `src/docs/chapters/chapter1.adoc´ is rendered.

To avoid trouble, the asciidoc docs suggest to use an absolute path for images. And this is wha docToolchain does.
It sets the `imagesdir` to `src/docs/images`. 
The `imageDirs` list in the config tells docToolchain to copy everything from these dirs to `src/docs/images`. This is a workaround to have iamges in other folders.

Does this help to understand the logic?

**Reply by ponchofiesta** [reply]
Thank you for clarifing. Does it mean, this can be solved only be placing images near to the document where it is used? And using one image in multiple documents in a folder hierarchy is impossible?

**Reply by rdmueller** [reply]
No, that's not what it means.
Here is my approach:

if I have a file like

    src/docs/arc42/chapters/chapter1.adoc

I place my images in

    src/docs/images/arc42

I then reference them as

    image::arc42/foo.png

to ensure that the editor displays the image, I add a line like this on top:

    ifndef::imagesdir[:imagesdir: ../../images]

This sets the relative `imagesdir` when it is not defined (which is the case in the editor). It is ignored when the `imagesdir` is already set to the absolute path (when docToolchain builds)

Does this help? Maybe we should transform this information in a tutorial...

**Reply by ponchofiesta** [reply]
Hmm, can't get it working like this. As I assume this workaround works with one image source per document only, I moved my images to a dir for this doc.

I have my images in `src/asciidoc/doc/product/Anwenderhandbuch/images`.
The Document is `src/asciidoc/doc/product/Anwenderhandbuch/Anwenderhandbuch.adoc`.
There I have set `ifndef::imagesdir[:imagesdir: images]`
This doc includes `include::Anwenderhandbuch_00-01-00_Titelseite.adoc[]`
And in Titelseite it includes the image `image::internal.svg[width=180]`. (I also tried PNG)
In docToolchainConfig.groovy this is configured:
```
imageDirs = [
    "src/asciidoc/doc/product/Anwenderhandbuch/images",
]
```

The preview in VSCode is working.
But doctoolchain throws:

```
* Where:
Script '/home/dtcuser/docToolchain/scripts/publishToConfluence.gradle' line: 73

* What went wrong:
Execution failed for task ':publishToConfluence'.
> java.io.FileNotFoundException: /project/build/html5/src/asciidoc/doc/product/Anwenderhandbuch/images/internal.svg (No such file or directory)
```

**Reply by rdmueller** [reply]
ah! you are working with publishToConfluence in a docker container.
I guess I have to give it a try myself. Can you setup a public repo where I can reproduce the problem? I guess we don't have to setup publishToConfluence in this repo. A working generateHTML would be enough.

**Reply by ponchofiesta** [reply]
Sorry, didn't say that. I created a MRE which causes the problem for me too:
https://github.com/ponchofiesta/doctoolchain-demo
docToolchainConfig.groovy needs to be adjusted by your environment.

I Try to publish it by running this:

```
docker run -u$(id -u):$(id -g) \
	--rm \
	-it \
	--entrypoint /bin/bash \
	-w /project \
	-v $(pwd):/project \
	XXXXXXXXXXXXXXXX/doctoolchain/doctoolchain:v3.4.0 \
	-c "doctoolchain . publishToConfluence  \
	-PmainConfigFile=docToolchainConfig.groovy \
	-PmavenRepository=XXXXXXXXX \
	-PmavenPassword="$ARTIFACTORY_TOKEN" \
	-PdepsMavenRepository=XXXXXXXXX \
	-PdepsMavenPassword="$ARTIFACTORY_TOKEN" \
	--info"
```

**Reply by rdmueller** [reply]
interesting to see that many people avoid using the dtcw script

**Reply by ponchofiesta** [reply]
I saw risks, e.g. internet access, internal repositories. If there is a docker image, in most cases I'd prefer it over such "local" solutions. IMO Docker provides all of the advantages, the script should do and it's an established tool.

**Reply by rdmueller** [reply]
Thanx for the Demo. Yes, there seems to be a bug!

**Reply by rdmueller** [reply]
I created two PRs for you. The second one restructures the whole repository to be able to render HTML, PDF and Site. With HTML working, publishToConfluence should also work.

There is a small workaround for the bug you found - the image folder has to be copied once.

So, in order to use publishToConfluence, do a `generateSite`, `cp` and then `publishToConfluence` should work.

**Reply by ponchofiesta** [reply]
Thank you very much! generateSite caused new problems. The command converted each file in the input directory, ran forever and it printed many extremely long paths. It blowed up the build directory. Looks like it did some strange recursion. But generateHTML worked.

Summary:
- All images are in one directory at the moment
- In each doc that uses an image, I put `ifndef::imagesdir[:imagesdir: ../../images]` on the top pointing to the correct relative images path.
- Run this to build and publish (for all 6 documents that are published in my project, only 3 of them do use images):
  ```bash
  # doctoolchain is an alias for the docker command
  doctoolchain generateHTML
  cp -r build/html5/images build/html5/product/Anforderungsdefinition/
  cp -r build/html5/images build/html5/product/Anwenderhandbuch/
  cp -r build/html5/images build/html5/projectManagement/Projekthandbuch/
  doctoolchain publishToConfluence
  ```

Not really great but it works for now.

**Reply by rdmueller** [reply]
"strange recursion"
when you set the inputPath to '.' it will crawl the whole project for asciidoc files and copy them to the build folder. This will result in some asciidoc files in the build folder after the first run. The second run will find those asciidoc file within the build folder and copy them deeper into the build folder...

That's why you should always set the inputPath to some folder. And remember, you can always safely delete the build folder
